# Indel-for-COVID-19
This reposity include codes for our manuscript "Identification of recurrent indels associated with fitness of SARS-CoV-2".
In this study, we developed a bioinformatics pipeline to systematically investigate insertions and deletions, dubbed by indels, occurring in over 9 million SARS-CoV-2 genomes and revealed 33,407,079 deletions and 2,046,903 insertions. We fork the repository from [BVAS](https://github.com/broadinstitute/bvas.git) and make our modifications to analyze Indels. 
Overall, we first refine the clusters based on Indel information. By employing the BVAS inference, several Indels are found to have comparable selection coefficient with dominant substitutions, as is shown in [Fig.1](https://user-images.githubusercontent.com/58204352/201807914-6d13e156-2225-43df-82fd-068fec989dd9.png), which implies the majority of studies for covid would suffer substantially from overlooking of Indels. ![Fig.1 Estimation of selection coefficients of substitutions](https://user-images.githubusercontent.com/58204352/201807914-6d13e156-2225-43df-82fd-068fec989dd9.png),
